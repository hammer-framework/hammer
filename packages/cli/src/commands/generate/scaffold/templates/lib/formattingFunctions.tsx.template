import React from 'react'

import humanize from 'humanize-string'

const MAX_STRING_LENGTH = 150

export const formatEnum = (values: string | string[] | null | undefined) => {
  if (values) {
    if (Array.isArray(values)) {
      const humanizedValues = values.map((value) => humanize(value))
      return humanizedValues.join(', ')
    } else {
      return humanize(values as string)
    }
  }

  return ''
}

export const truncate = (value: string | number) => {
  const output = value?.toString()
  if (output?.length > MAX_STRING_LENGTH) {
    return output.substring(0, MAX_STRING_LENGTH) + '...'
  }
  return output ?? ''
}

export const jsonTruncate = (obj: unknown) => {
  return truncate(JSON.stringify(obj, null, 2))
}

export const timeTag = (datetime?: string) => {
  return datetime ? (
    <time dateTime={datetime} title={datetime}>
      {new Date(datetime).toUTCString()}
    </time>
  ) : (
    ''
  )
}

export const checkboxInputTag = (checked: boolean) => {
  return <input type="checkbox" checked={checked} disabled />
}
