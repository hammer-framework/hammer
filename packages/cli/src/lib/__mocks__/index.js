export const getPaths = () => MOCK_PATHS

export const MOCK_PATHS = {
  base: 'mock_basedir',
  generated: {
    base: '.mock_redwood',
    schema: '.mock_redwood/schema.graphql',
    types: {
      includes: '.mock_redwood/types/includes',
      mirror: '.mock_redwood/types/mirror',
    },
    prebuild: '.mock_redwood/prebuild',
  },
  scripts: 'mock_scripts_dir',
  api: {
    base: 'mock_api',
    dataMigrations: 'mock_dataMigrations',
    db: 'mock_db_dir',
    dbSchema: 'mock_db_schema_dir',
    functions: 'fmock_unctions_dir',
    graphql: 'mock_graphql_dir',
    lib: 'mock_lib_dir',
    generators: 'mock_generators_dir',
    config: 'mock_config_dir',
    services: 'mock_services_dir',
    directives: 'mock_directives_dir',
    src: 'mock_src_dir',
    dist: 'mock_api/dist',
    types: 'mock_api/types',
    models: 'mock_models_dir',
  },
  web: {
    routes: ['mock_mock/routes'],
    base: 'mock_web',
    pages: 'mock_pages_dir',
    components: 'mock_components_dir',
    layouts: 'mock_layouts_dir',
    src: 'mock_src_dir',
    generators: 'mock_generators_dir',
    app: 'mock_app_dir',
    index: 'mock_index_dir',
    config: 'mock_config_dir',
    webpack: 'mock_webpack_dir',
    postcss: 'mock_postcss_dir',
    storybookConfig: 'mock_storybookConfig_dir',
    storybookPreviewConfig: 'mock_storybookPreviewConfig_dir',
    storybookManagerConfig: 'mock_storybookManagerConfig_dir',
    dist: 'mock_web/dist',
    types: 'mock_web/types',
  },
}
