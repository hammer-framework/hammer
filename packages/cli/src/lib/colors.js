import chalk from 'chalk'

/**
 * To keep a consistent color/style palette between cli packages, such as
 * @redwood/cli and @redwood/create-redwood-app, please keep them compatible
 * with one and another. We'll might split up and refactor these into a
 * separate package when there is a strong motivation behind it.
 *
 * Current files:
 *
 * - packages/cli/src/lib/colors.js (this file)
 * - packages/create-redwood-app/src/create-redwood-app.js
 */
export default {
  error: chalk.bold.red,
  warning: chalk.keyword('orange'),
  success: chalk.greenBright,
  info: chalk.grey,

  bold: chalk.bold,
  underline: chalk.underline,

  green: chalk.green,

  header: chalk.bold.underline.hex('#e8e8e8'),
  cmd: chalk.hex('#808080'),
  redwood: chalk.hex('#ff845e'),
  love: chalk.redBright,
}
