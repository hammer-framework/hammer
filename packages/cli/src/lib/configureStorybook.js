import fs from 'fs-extra'

import { getPaths } from '.'

/**
 * Configure Storybook for the given template by creating a custom preview config
 */
export default function configureStorybook({ force }, newStorybookPreview) {
  const storybookPreviewConfigPath = getPaths().web.storybookPreviewConfig

  let storybookPreviewConfig
  /**
   *  Check if storybookPreviewConfigPath already exists.
   *  Merge both files if it does.
   *  By removing import react and export decorator from new config
   *  And put new config inside current config after last import
   */
  if (fs.existsSync(storybookPreviewConfigPath)) {
    if (force) {
      fs.unlinkSync(storybookPreviewConfigPath)
      storybookPreviewConfig = newStorybookPreview
    } else {
      const currentConfig = fs
        .readFileSync(storybookPreviewConfigPath)
        .toString()

      const newDecoratorsName = newStorybookPreview.match(
        /export const decorators = \[(.*?)\]/
      )[1]

      const currentDecoratorsName = currentConfig.match(
        /export const decorators = \[(.*?)\]/
      )[1]

      const decoratorsExport = `export const decorators = [${currentDecoratorsName}, ${newDecoratorsName}]`

      const insideNewStorybookConfigWithoutReactAndDecoration =
        newStorybookPreview
          .replace(/import \* as React from 'react'/, '')
          .replace(/export const decorators = .*/, '')

      const currentConfigWithoutDecoration = currentConfig.replace(
        /export const decorators = .*/,
        ''
      )

      const reversedCurrentConfig = currentConfigWithoutDecoration
        .split('\n')
        .reverse()

      const indexOfLastImport = reversedCurrentConfig.findIndex((value) =>
        /^import /.test(value)
      )
      reversedCurrentConfig.splice(
        indexOfLastImport,
        0,
        insideNewStorybookConfigWithoutReactAndDecoration
      )
      storybookPreviewConfig =
        reversedCurrentConfig.reverse().join(`\n`) +
        `\n` +
        currentConfig +
        `\n` +
        decoratorsExport
    }
  } else {
    storybookPreviewConfig = newStorybookPreview
  }

  fs.outputFileSync(storybookPreviewConfigPath, storybookPreviewConfig)
}
