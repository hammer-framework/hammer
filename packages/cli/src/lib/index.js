import fs from 'fs'
import path from 'path'

import lodash from 'lodash/string'
import camelcase from 'camelcase'
import pascalcase from 'pascalcase'
import pluralize from 'pluralize'
import { paramCase } from 'param-case'
import { getDMMF } from '@prisma/sdk'
import { getPaths as getRedwoodPaths } from '@redwoodjs/internal'
import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import c from 'src/lib/colors'

export const asyncForEach = async (array, callback) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

/**
 * Returns the database schema for the given `name` database table parsed from
 * the schema.prisma of the target applicaiton. If no `name` is given then the
 * entire schema is returned.
 */
export const getSchema = async (name) => {
  const schemaPath = path.join(getPaths().api.db, 'schema.prisma')
  const metadata = await getDMMF({
    datamodel: readFile(schemaPath.toString()),
  })

  if (name) {
    const model = metadata.datamodel.models.find((model) => {
      return model.name === name
    })

    if (model) {
      return model
    } else {
      throw `No schema definition found for \`${name}\``
    }
  }

  return metadata.datamodel
}

/**
 * Returns variants of the passed `name` for usage in templates. If the given
 * name was "fooBar" then these would be:

 * pascalName: FooBar
 * singularPascalName: FooBar
 * pluralPascalName: FooBars
 * singularCamelName: fooBar
 * pluralCamelName: fooBars
 * singularParamName: foo-bar
 * pluralParamName: foo-bars
*/
const nameVariants = (name) => {
  const normalizedName = pascalcase(pluralize.singular(name))

  return {
    pascalName: pascalcase(name),
    camelName: camelcase(name),
    singularPascalName: normalizedName,
    pluralPascalName: pluralize(normalizedName),
    singularCamelName: camelcase(normalizedName),
    pluralCamelName: camelcase(pluralize(normalizedName)),
    singularParamName: paramCase(normalizedName),
    pluralParamName: paramCase(pluralize(normalizedName)),
  }
}

export const templateRoot = path.resolve(
  __dirname,
  '../commands/Generate/templates'
)

export const generateTemplate = (templateFilename, { name, ...rest }) => {
  const templatePath = path.join(templateRoot, templateFilename)
  const template = lodash.template(readFile(templatePath).toString())
  return template({
    name,
    ...nameVariants(name),
    ...rest,
  })
}

export const readFile = (target) => fs.readFileSync(target)

export const writeFile = async (
  target,
  contents,
  { overwriteExisting = false } = {}
) => {
  if (!overwriteExisting && fs.existsSync(target)) {
    throw new Error(`${target} already exists.`)
  }

  const filename = path.basename(target)
  const targetDir = target.replace(filename, '')
  fs.mkdirSync(targetDir, { recursive: true })
  fs.writeFileSync(target, contents)
}

export const bytes = (contents) => Buffer.byteLength(contents, 'utf8')

/**
 * This wraps the core version of getPaths into something that catches the exception
 * and displays a helpful error message.
 */
export const getPaths = () => {
  try {
    return getRedwoodPaths()
  } catch (e) {
    console.log(c.error(e.message))
    process.exit(0)
  }
}

/**
 * Creates a list of tasks that write files to the disk.
 *
 * @param files - {[filepath]: contents}
 */
export const writeFilesTask = (files, options) => {
  const { base } = getPaths()
  return new Listr(
    Object.keys(files).map((file) => {
      const contents = files[file]
      return {
        title: `Writing \`./${path.relative(base, file)}\`...`,
        task: () => writeFile(file, contents, options),
      }
    })
  )
}

/**
 * Update the project's routes file.
 */
export const addRoutesToRouterTask = (routes) => {
  const redwoodPaths = getPaths()
  const routesContent = readFile(redwoodPaths.web.routes).toString()
  const newRoutesContent = routes.reduce((content, route) => {
    return content.replace(/(\s*)\<Router\>/, `$1<Router>$1  ${route}`)
  }, routesContent)
  writeFile(redwoodPaths.web.routes, newRoutesContent, {
    overwriteExisting: true,
  })
}

export const runCommandTask = async (commands, { verbose }) => {
  const tasks = new Listr(
    commands.map(({ title, cmd, args, opts = {} }) => ({
      title,
      task: async () => {
        return execa(cmd, args, {
          shell: true,
          cwd: `${getPaths().base}/api`,
          stdio: verbose ? 'inherit' : 'pipe',
          extendEnv: true,
          cleanup: true,
          ...opts,
        })
      },
    })),
    {
      renderer: verbose && VerboseRenderer,
      dateFormat: false,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}

export * from './generateTempSchema'
