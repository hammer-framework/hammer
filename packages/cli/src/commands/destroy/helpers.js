import Listr from 'listr'

import { deleteFilesTask } from 'src/lib'
import c from 'src/lib/colors'

const tasks = ({ componentName, filesFn, name }) =>
  new Listr(
    [
      {
        title: `Destroying ${componentName} files...`,
        task: async () => {
          const f = await filesFn({ name })
          return deleteFilesTask(f)
        },
      },
    ],
    { collapse: false, exitOnError: true }
  )

export const createYargsForComponentDestroy = ({ componentName, filesFn }) => {
  return {
    command: `${componentName} <name>`,
    desc: `Destroy a ${componentName} component.`,
    handler: async ({ name }) => {
      const t = tasks({ componentName, filesFn, name })

      try {
        await t.run()
      } catch (e) {
        console.log(c.error(e.message))
      }
    },
    tasks,
  }
}
