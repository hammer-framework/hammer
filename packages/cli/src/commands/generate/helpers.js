import path from 'path'

import camelcase from 'camelcase'
import pluralize from 'pluralize'
import Listr from 'listr'
import pascalcase from 'pascalcase'
import { paramCase } from 'param-case'

import { generateTemplate, getPaths, writeFilesTask } from 'src/lib'
import c from 'src/lib/colors'

/**
 * Reduces boilerplate for generating an output path and content to write to disk
 * for a component.
 */
// TODO: Make this read all the files in a template directory instead of
// manually passing in each file.
export const templateForComponentFile = ({
  name,
  suffix = '',
  extension = '.js',
  webPathSection,
  apiPathSection,
  generator,
  templatePath,
  templateVars,
  componentName,
}) => {
  const basePath = webPathSection
    ? getPaths().web[webPathSection]
    : getPaths().api[apiPathSection]
  const outputComponentName =
    componentName || pascalcase(paramCase(name)) + suffix
  const outputPath = path.join(
    basePath,
    outputComponentName,
    outputComponentName + extension
  )
  const content = generateTemplate(
    path.join(generator, 'templates', templatePath),
    {
      name,
      outputPath: `./${path.relative(getPaths().base, outputPath)}`,
      ...templateVars,
    }
  )
  return [outputPath, content]
}

/**
 * Creates a route path, either returning the existing path if passed, otherwise
 * creates one based on the name
 */
export const pathName = (path, name) => {
  return path ?? `/${paramCase(name)}`
}

/**
 * Reduces boilerplate for creating a yargs handler that writes a component to a
 * location.
 */
// TODO: Make this work for all files, not just components.
export const createYargsForComponentGeneration = ({
  componentName,
  filesFn,
}) => {
  return {
    command: `${componentName} <name>`,
    desc: `Generate a ${componentName} component.`,
    builder: { force: { type: 'boolean', default: false } },
    handler: async ({ force, ...rest }) => {
      const tasks = new Listr(
        [
          {
            title: `Generating ${componentName} files...`,
            task: async () => {
              const f = await filesFn(rest)
              return writeFilesTask(f, { overwriteExisting: force })
            },
          },
        ],
        { collapse: false, exitOnError: true }
      )

      try {
        await tasks.run()
      } catch (e) {
        console.log(c.error(e.message))
      }
    },
  }
}

// Returns all relations to other models
export const relationsForModel = (model) => {
  return model.fields
    .filter((f) => f.relationName)
    .map((field) => {
      const relationName = camelcase(field.type)
      return field.isList ? pluralize(relationName) : relationName
    })
}

// Returns only relations that are of datatype Int
export const intForeignKeysForModel = (model) => {
  return model.fields
    .filter((f) => f.name.match(/Id$/) && f.type === 'Int')
    .map((f) => f.name)
}
