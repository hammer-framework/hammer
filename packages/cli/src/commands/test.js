import execa from 'execa'
import Listr from 'listr'
import VerboseRenderer from 'listr-verbose-renderer'

import { getPaths } from 'src/lib'
import c from 'src/lib/colors'

export const command = 'test [app..]'
export const desc = 'Run Jest tests for api and web.'
export const builder = {
  app: { choices: ['api', 'web'], default: ['api', 'web'] },
}

export const handler = async ({ app }) => {
  const { base: BASE_DIR } = getPaths()
  const execCommands = {
    api: {
      cwd: `${BASE_DIR}/api`,
      cmd: 'yarn jest',
      args: [
        '--passWithNoTests',
        '--config ../node_modules/@redwoodjs/core/config/jest.config.api.js',
      ],
    },
    web: {
      cwd: `${BASE_DIR}/web`,
      cmd: 'yarn jest',
      args: [
        '--passWithNoTests',
        '--config ../node_modules/@redwoodjs/core/config/jest.config.web.js',
      ],
    },
  }

  const tasks = new Listr(
    app.map((appName) => {
      const { cmd, args, cwd } = execCommands[appName]
      return {
        title: `Running '${appName}' jest tests`,
        task: () => {
          return execa(cmd, args, {
            stdio: 'inherit',
            shell: true,
            cwd,
          })
        },
      }
    }),
    {
      renderer: VerboseRenderer,
    }
  )

  try {
    await tasks.run()
  } catch (e) {
    console.log(c.error(e.message))
  }
}
