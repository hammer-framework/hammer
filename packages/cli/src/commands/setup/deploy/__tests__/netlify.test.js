// Automock fs using ../..../__mocks__/fs
jest.mock('fs')

import path from 'path'

import fs from 'fs-extra'

import { getPaths } from '../../../../lib'
import { updateApiURLTask } from '../helpers'
// Mock telemetry and other things
import '../../../../lib/test'

jest.mock('../../../../lib', () => {
  const path = jest.requireActual('path')

  const { printSetupNotes } = jest.requireActual('../../../../lib')

  return {
    printSetupNotes,
    getPaths: () => {
      return {
        base: path.resolve(
          path.join(
            __dirname,
            '../../../../../../../__fixtures__/example-todo-main'
          )
        ),
      }
    },
    getConfig: () => ({
      web: {
        port: 8910,
      },
    }),
    writeFilesTask: (fileNameToContentMap) => {
      const keys = Object.keys(fileNameToContentMap)
      expect(keys.length).toBe(1)
      // Need to escape path.sep on Windows, otherwise the backslash (that
      // path.sep is on Windows) together with the 'n' in "netlify" will be
      // interpreted as a new-line. And need to use double backslashes, so
      // that one "survives" into the regexp
      expect(keys[0]).toMatch(new RegExp(`\\${path.sep}netlify.toml$`))
      expect(fileNameToContentMap[keys[0]]).toMatchSnapshot()
    },
  }
})

const REDWOOD_TOML_PATH = path.join(getPaths().base, 'redwood.toml')

beforeEach(() => {
  fs.__setMockFiles({
    [REDWOOD_TOML_PATH]: `[web]
  title = "Redwood App"
  port = 8910
  apiUrl = "/.redwood/functions" # you can customize graphql and dbAuth urls individually too: see https://redwoodjs.com/docs/app-configuration-redwood-toml#api-paths
  includeEnvironmentVariables = [] # any ENV vars that should be available to the web side, see https://redwoodjs.com/docs/environment-variables#web
[api]
  port = 8911
[browser]
  open = true
`,
  })
})

describe('netlify', () => {
  it('should call the handler without error', async () => {
    const netlify = require('../providers/netlify')
    expect(async () => await netlify.handler({ force: true })).not.toThrow()
  })

  it('Should update redwood.toml apiUrl', () => {
    updateApiURLTask('/.netlify/functions').task()

    expect(fs.readFileSync(REDWOOD_TOML_PATH)).toMatch(
      /apiUrl = "\/.netlify\/functions"/
    )
  })

  it('should add netlify.toml', async () => {
    const netlify = require('../providers/netlify')
    await netlify.handler({ force: true })
    // Will be verified by a snapshot up above in the mocked `writeFilesTask`
  })
})
