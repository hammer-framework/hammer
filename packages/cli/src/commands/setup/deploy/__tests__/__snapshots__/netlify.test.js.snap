// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`netlify should add netlify.toml 1`] = `
"[build]
  command = "yarn rw deploy netlify"
  publish = "web/dist"
  functions = "api/dist/functions"

[dev]
  # To use [Netlify Dev](https://www.netlify.com/products/dev/),
  # install netlify-cli from https://docs.netlify.com/cli/get-started/#installation
  # and then use netlify link https://docs.netlify.com/cli/get-started/#link-and-unlink-sites
  # to connect your local project to a site already on Netlify
  # then run netlify dev and our app will be accessible on the port specified below
  framework = "redwoodjs"
  # Set targetPort to the [web] side port as defined in redwood.toml
  targetPort = 8910
  # Point your browser to this port to access your RedwoodJS app
  port = 8888

[[redirects]]
  from = "/*"
  to = "/200.html"
  status = 200
"
`;

exports[`netlify should call the handler without error 1`] = `
"[build]
  command = "yarn rw deploy netlify"
  publish = "web/dist"
  functions = "api/dist/functions"

[dev]
  # To use [Netlify Dev](https://www.netlify.com/products/dev/),
  # install netlify-cli from https://docs.netlify.com/cli/get-started/#installation
  # and then use netlify link https://docs.netlify.com/cli/get-started/#link-and-unlink-sites
  # to connect your local project to a site already on Netlify
  # then run netlify dev and our app will be accessible on the port specified below
  framework = "redwoodjs"
  # Set targetPort to the [web] side port as defined in redwood.toml
  targetPort = 8910
  # Point your browser to this port to access your RedwoodJS app
  port = 8888

[[redirects]]
  from = "/*"
  to = "/200.html"
  status = 200
"
`;
